examples/threading/threading.c

```diff
@@ -11,23 +11,70 @@
 void* threadfunc(void* thread_param)
 {
 
-    // TODO: wait, obtain mutex, wait, release mutex as described by thread_data structure
-    // hint: use a cast like the one below to obtain thread arguments from your parameter
-    //struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+    struct thread_data *prams = (struct thread_data *) thread_param;
+    prams->thread_complete_success = true;
+
+
+    // sleep before acquiring lock
+    if (usleep(prams->wait_time_for_lock * 1000) != 0) {
+	ERROR_LOG("\r\n Failed to sleep before acquiring locks");
+	prams->thread_complete_success = false;
+	return thread_param;
+    }
+
+    //acquire lock 
+    if (pthread_mutex_lock(prams->lock) != 0) {
+    	ERROR_LOG("\r\n Failed to acquire lock");
+	prams->thread_complete_success = false;
+        return thread_param;
+
+    }
+
+    //sleep before releasing lock
+    if (usleep(prams->wait_time_to_release_lock * 1000) != 0) {
+	ERROR_LOG("\r\n Failed to sleep after acquiring locks");
+	pthread_mutex_unlock(prams->lock);
+	prams->thread_complete_success = false;
+        return thread_param;
```

Comment:
@maheshchnt nice work remembering to free in this case.  Think about having one point for cleanup after unlock rather than duplicating cleanup logic in every failure case and doing returns in your error handling block.